common/predictive-text/worker/model-compositor.ts

```diff
@@ -58,15 +63,30 @@ class ModelCompositor {
       } else if(this.isBackspace(transform) && !allowBksp) {
         continue;
       }
+
+      let preserveWhitespace: boolean = false;
+      if(this.isWhitespace(transform)) {
+        // Detect start of new word; prevent whitespace loss here.
+        let postContext = models.applyTransform(transform, context);
+        preserveWhitespace = (this.lexicalModel.wordbreak(postContext) == '');
+      }
+
       let distribution = this.lexicalModel.predict(transform, context);
 
+      let mc = this;
       distribution.forEach(function(pair: ProbabilityMass<Suggestion>) {
         // Let's not rely on the model to copy transform IDs.
         // Only bother is there IS an ID to copy.
         if(transform.id !== undefined) {
           pair.sample.transformId = transform.id;
         }
 
+        // Prepends the original whitespace, ensuring it is preserved if
+        // the suggestion is accepted.
+        if(preserveWhitespace) {
+          models.prependTransform(pair.sample.transform, transform);
+        }
```

Comment:
:+1:

However, implicitly this means that the underlying models is responsible for determining whether to insert spaces. I'm not sure this is the correct place to put it. Just a thinking out loud right now...

Reply:
Keep in mind that `models` here is the namespace, not the actual lexical model.

This line is what allows us to preserve spaces and line breaks when taking a null-context suggestion.  Ideally, if that's supposed to be a zero-width space for 'whitespace', the keyboard actually forwarded the ZWSP as the transform, which this would then also preserve.  (To be safe, I've added `\u200b' to our whitespace check accordingly.)

Reply:
Ah, okay! Thanks for clearing that up!